### PR TITLE
Avoid treating `.whl` sources as source distributions

### PR DESCRIPTION
**Pull request by charliermarsh**
## Summary

The error messages here are incorrect.

Closes https://github.com/astral-sh/uv/issues/7284.

### CODE REVIEW

File: crates/uv/tests/sync.rs [hunk 1]
```diff
@@ -2161,3 +2161,92 @@ fn no_build() -> Result<()> {
...
+
+    uv_snapshot!(context.filters(), context.sync(), @r###"
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    warning: Failed to validate existing lockfile: distribution cffi==1.17.1 @ path+cffi-1.17.1-cp310-cp310-macosx_11_0_arm64.whl can't be installed because it doesn't have a source distribution or wheel for the current platform
+    Resolved 3 packages in [TIME]
+    error: distribution cffi==1.17.1 @ path+cffi-1.17.1-cp310-cp310-macosx_11_0_arm64.whl can't be installed because it doesn't have a source distribution or wheel for the current platform
```
Comment (zanieb) [inline crates/uv/tests/sync.rs:2248]: The wheel can't be installed because it doesn't have a source distribution for the current platform?

Comment (zanieb) [reply crates/uv/tests/sync.rs:2248]: Is this message generally correct and you're using these for test cases?

Should we open an issue to add a special case for direct URLs to wheels or sdists?

File: crates/uv/tests/sync.rs [hunk 2]
```diff
@@ -2161,3 +2161,92 @@ fn no_build() -> Result<()> {
...
+    Resolved 3 packages in [TIME]
+    "###);
+
+    uv_snapshot!(context.filters(), context.sync(), @r###"
+    success: false
+    exit_code: 2
+    ----- stdout -----
+
+    ----- stderr -----
+    warning: Failed to validate existing lockfile: distribution cffi==1.17.1 @ path+cffi-1.17.1-cp310-cp310-macosx_11_0_arm64.whl can't be installed because it doesn't have a source distribution or wheel for the current platform
```
Comment (zanieb) [inline crates/uv/tests/sync.rs:2246]: Kind of annoying we show a warning that matches the error message

Comment (charliermarsh) [reply crates/uv/tests/sync.rs:2246]: This is a bug actually

File: crates/uv/tests/sync.rs (hunk 1 above)
Comment (charliermarsh) [reply crates/uv/tests/sync.rs:2248]: Yeah it's generally correct, you're right that it's off here though. I'll fix it up in a separate PR.

**Pull request merged**